### PR TITLE
Its maps development

### PR DESCRIPTION
**Pull request by adfrawley**
Merged my adfrawley/coresoftware "ITS_MAPS_development" branch with the master repository. This branch contained the code to construct the sPHENIX tracker 3-layer inner barrel from ITS MAPS staves, using the stave geometry imported from the ALICE simulations software. 

I also made a macro that constructs the sPHENIX tracker as (3 layers of maps ladders + 4 layers of silicon strips + 60 layers of TPC gas). This will separately be added to the macros directory in the master repository. NOTE: this new macro sets the last tracking layer of the TPC at 78 cm, instead of the previously used 80 cm.

I ran some tracking tests using single pions to compare performance with the MAPS ladders and with the cylinder cell geometry. The cylinder cell geometry simulations were made with the outer radius of the TPC at 80 cm, and the maps ladders simulations were with the TPC outer radius at 78 cm. This explains the 2% difference in pT resolution between cylinder maps and ladder maps cases in the plots. 

The results of the tracking tests are shown here:

1) Comparison of performance for pT resolution and dca2d resolution with cylinder maps or ladder maps + TPC.
[compare_cylinder_maps_ladder_maps_tpc_ptres.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/623481/compare_cylinder_maps_ladder_maps_tpc_ptres.pdf)
[compare_cylinder_maps_ladder_maps_tpc_dca2d.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/623487/compare_cylinder_maps_ladder_maps_tpc_dca2d.pdf)

2) Comparison of pT and dca2d performance with cylinder maps or ladder maps + intt + TPC.
[compare_cylinder_maps_ladder_maps_intt_tpc_ptres.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/623485/compare_cylinder_maps_ladder_maps_intt_tpc_ptres.pdf)
[compare_cylinder_maps_ladder_maps_intt_tpc_dca2d.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/623488/compare_cylinder_maps_ladder_maps_intt_tpc_dca2d.pdf)

### CODE REVIEW

File: simulation/g4simulation/g4detectors/PHG4MapsCellReco.cc [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+		  celllist[key]->set_half_stave_index(half_stave_number);
+		  celllist[key]->set_module_index(module_number);
+		  celllist[key]->set_chip_index(chip_number);
+		  celllist[key]->set_pixel_index(pixel_number);
+		  
+		  celllist[key]->set_phibin(vxbin[i1]);
+		  celllist[key]->set_zbin(vzbin[i1]);
+		  
+		  double edep;
+		  if(trklen > 0.0)
```
Comment (blackcathj) [inline simulation/g4simulation/g4detectors/PHG4MapsCellReco.cc]: Valgrind complained on this line:  ```trklen``` was declared earlier without a default value. And in line 240 ```if(pixel_number_out != pixel_number_in)```, if the evaluation is false, then ```trklen``` would not be assigned. Please double check this. Maybe the solution is to init  ```trklen=0``` in line 235.

File: simulation/g4simulation/g4detectors/PHG4MapsDetector.cc [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+      cout << endl;
+    }
+
+
+  //===================================
+  // Import the stave physical volume here
+  //===================================
+
+  // import the staves from the gemetry file
+  G4GDMLReadStructure * reader = new G4GDMLReadStructure();
```
Comment (blackcathj) [inline simulation/g4simulation/g4detectors/PHG4MapsDetector.cc]: This leads to a memory leakage as apparently ```G4GDMLParser``` later does not delete this reader. I will fix it in a later new pull request.

**Comment by blackcathj** [comment]
Just to show the ITS geometry importing and layout works, few event display got during the testing:

ladders

ladders2

ladders3

ladders4

**Comment by sourav-tarafdar** [comment]
Not that much of difference in tracking performance between cylindrical and ladder geometry as is shown in the plot. The tracking performance is with pure pions for MAPS+IT+TPC configuration with 60 gas layers of TPC.
[ladder_vs_cyl_maps+it+tpc_purepions_v2.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/634598/ladder_vs_cyl_maps.it.tpc_purepions_v2.pdf)

**Comment by sourav-tarafdar** [comment]
As per discussion during sPHENIX simulation meeting on 12/06/2016 here is the new set of tracking performance comparison plots between ladder geometry libraries and cylindrical geometry libraries using cylindrical geometry macros for both cases. The current set of plot is being made using MAPS+IT+TPC tracking configuration for pure pions. In previous set of plot posted and discussed in the sPHENIX simulation meeting, concern was raised regarding slight discrepancy in tracking efficiency. Suggestion was to remake the cylindrical geometry tracking performance plots using latest nightly build libraries rather than using tracking performance plot made few months ago for comparison purpose. Redoing the exercise removed the discrepancy in track efficiency comparison plot. This exercise validates that using ladder geometry libraries doesn't break existing macros or tracking performance. 
[maps_it_tpc_ladder_pull218.pdf](https://github.com/sPHENIX-Collaboration/coresoftware/files/635872/maps_it_tpc_ladder_pull218.pdf)

**Comment by blackcathj** [comment]
Thanks to @sourav-tarafdar for the cross check and confirmation. Also checked the cell-ID for MAPS should be unique in order to trace truth information from ```SvtxHit``` -> ```PHG4CylinderCell_MAPS```. Merging this pull request

**Pull request merged**